STYLEGUIDE.md

```diff
@@ -0,0 +1,367 @@
+- [CSS-in-JS](#CSS-in-JS)
+- [File Structure](https://github.com/mongodb/leafygreen-ui/blob/main/stories/Folder-Structure.stories.mdx)
+- [References](#References)
+
+## <a id="To-Contribute">To Contribute</a>
+
+To propose updates to this style guide, please make the relevant changes in a branch and submit a PR. We review these PRs synchronously in our weekly Team Code Review and collectively gavel on new standards there.
+
+### Format Contributions
+
+### Prefer [Rule]
+
+#### Why
+
+_Description_
+
+#### Prefer
+
+_Example_
+
+#### Avoid
+
+_Example_
+
+---
+
+## <a id="Typing">Typing</a>
+
+### Prefer constants and types over enums.
+
+#### Why
+
+Ensures that TypeScript doesn't prevent people from using the values directly rather than consuming the enum.
+
+#### Prefer
+
+```typescript
+export const someConstant = {
+  optionOne = 'option 1',
+  optionTwo = 'option 2',
+} as const;
+```
+
+#### Avoid
+
+```typescript
+enum someConstant = {
+ optionOne: 'option 1',
+ optionTwo: 'option 2',
+}
+```
+
+---
+
+### When creating TS entities, place them at the top of the file, above JS
+
+---
+
+## <a id="Functions">Functions</a>
+
+### Prefer ternaries over simple `if/else` blocks
+
+#### Why
+
+Keeps code cleaner and easier to read.
+
+#### Prefer
+
+```typescript
+return isBlah ? a : b;
+```
+
+#### Avoid
+
+```typescript
+if (isBlah) {
+  return a;
+} else {
+  return b;
+}
+```
+
+---
+
+### All new (exported) functions should have a JSDoc comment explaining functionality.
+
+#### Why
+
+Documenting functions with [JSDoc](https://jsdoc.app/about-getting-started.html) helps PR reviewers better understand the intent of the function and also documents its use for other engineers using the code base. Better yet, [VS Code will automatically show in-line descriptions of a function](https://code.visualstudio.com/docs/languages/javascript#_jsdoc-support) when a user invokes that function in their editor.
+
+JSDocs should also clearly mention any major assumptions made about the function's input and output that may not be immediately obvious to an engineer unfamiliar with this part of the code base.
+
+#### Prefer
+
+```typescript
+/**
+ * Get the next applicable instance for upgrade. The function handles regions and cloud providers
+ * but does NOT work for Serverless clusters
+ * @param clusterDescription the backbone cluster description model
+ * @param providerOptionsModel the backbone provider options model
+ * @returns the next upgrade instance for this cluster
+ */
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+#### Avoid
+
+```typescript
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+---
+
+### Prefer creating the function in the highest scope, and passing in all required data.
+
+#### Why
+
+Keeps functions pure.
```

Comment:
Can also consider creating utility closures (e.g. https://github.com/mongodb/leafygreen-ui/blob/943e4191dabb74acf403b2189327216085cc3e81/packages/brand-shape/src/utils.ts#L15)

Reply:
What do you mean here?

Reply:
Sorry that was pretty unclear. Instead of defining a bunch of utilities that require component context, you can create a single utility function that accepts some component state as args, and returns a bunch of utilities with the component state in the closure. Not always useful, so it's fine to leave it out